Add a doc entry for setting up an OxCaml project with dune pkg (#12780)
ocaml/dune: doc/tutorials/dune-package-management/oxcaml.md

[OxCaml](https://oxcaml.org/) is a compiler branch with fast moving set of
extensions to OCaml. Some tweaks are required to set up a project that compiles
with OxCaml when using dune package management.

## Setting up the `dune-workspace`

OxCaml has a custom `opam-repository` that provides compatible dependencies.
This is specified in the `dune-workspace`.

```{code-block} dune
(lang dune 3.20)

(pkg enabled)

(repository
 (name oxcaml)
 (url git+https://github.com/oxcaml/opam-repository))

(lock_dir
 (repositories overlay oxcaml upstream))
```

In the `dune-project`, we also need to add a constraint to depend on the OxCaml
compiler. Otherwise, the solver may pick up the latest upstream compiler, which
does not support the OxCaml extensions.

```{code-block} dune
(lang dune 3.20)

(package
 (name hello-oxcaml)
 (depends
  (ocaml-variants
   (= 5.2.0+ox))))
```

Let's add a test program that uses OxCaml syntax.

::::{dropdown} `main.ml`
:icon: file-code

:::{literalinclude} oxcaml/main.ml
:language: dune
:::

::::

And to build it, we add a `dune` file.

```{code-block} dune
(executable
 (public_name main))
```

## Building the project

> As usual, let's build the project with:

Reviewer: Let's not assume users are used to the usual way of doing things here. Let's instead say, we can now build and execute our binary compiled with `OxCaml`. We use the `dune exec` command which will first build all the packages we need, including the OxCaml compiler and then builds our binary and runs it.

You should probably mention the passage of time here, since this step takes a while and dune atm doesn't give too much feedback.

```
$ dune exec ./main.exe
...
43
```
We have successfully built the project with the OxCaml compiler and run the executable.

## Developer tools

Note that OxCaml provides its own forks of the developer tools
`ocaml-lsp-server`, `ocamlformat`, etc. Make sure to include constraints for the
tools in the `dune-workspace` if you intend to use them for your development
workflows.

For `ocamlformat` it would look something like this:

```{code-block} dune

> (lang dune 3.20)

Reviewer: lang 3.21

(pkg enabled)

(repository
 (name oxcaml)
 (url "git+https://github.com/oxcaml/opam-repository"))

(pin
 (name ocamlbuild)

>  (url "git+https://github.com/Sudha247/ocamlbuild#oxcaml+dune")

Author: This points to my branch, I'm not sure it's a good idea for the docs - but it's also currently the working version which I intend to upstream to the oxcaml fork.

Reviewer 2: We just chatted: this is fine for now, let's just be sure to fix this once the oxcaml opam overlay is updated. We create an issue to track the need for that so we don't forget?

 (package
   (name ocamlbuild)
   (version 0.15.0+ox)))

(pin
 (name odoc-parser)
 (url "git+https://github.com/Sudha247/odoc#replace-symlinks")
 (package
   (name odoc-parser)
   (version 3.1.0+ox)))

(lock_dir

>  (path "dune.lock")

Reviewer 3: This path is probably optional.

 (repositories :standard oxcaml))

(lock_dir

>  (path "dev-tools.locks/ocamlformat")

Reviewer 3: This path needs to be adjusted now that #12671 is merged, it should be `"_build/.dev-tools.locks/ocamlformat"`

> # Example: Setting up an OxCaml project

Reviewer 2: ```suggestion
# Setting up an OxCaml project
```

As an example to adding custom repositories and pinning projects, let's look at
setting up a basic OxCaml project. [OxCaml](https://oxcaml.org/) is a compiler

> branch with fast moving set of extenstions to OCaml. Some tweaks are required to

Reviewer 2: ```suggestion
branch with a fast moving set of extensions to OCaml. Some tweaks are required to
```

> set up an OxCaml project with dune package management.

Reviewer 3: ```suggestion
set up an OxCaml project with Dune package management.
```

> ## Setting up the workspace

Reviewer 2: ```suggestion
## Setting up the `dune-workspace`
```

OxCaml has a custom `opam-repository` that we need in order to pull in its

> dependencies. This is specified in the `dune-workspace`.

Reviewer 2: ```suggestion
OxCaml has a custom `opam-repository` that provides compatible
dependencies. This is specified in the `dune-workspace`.
```

::::{dropdown} `dune-workspace`
:icon: file-code

:::{literalinclude} oxcaml/dune-workspace
:language: dune
:emphasize-lines: 3-5
:::

::::

In the `dune-project`, we also need to add a constraint to depend on the OxCaml
compiler. Otherwise, the solver may pick up the latest upstream compiler, which
does not support the OxCaml extensions.

::::{dropdown} `dune-project`
:icon: file-code

:::{literalinclude} oxcaml/dune-project
:language: dune
:::

::::

Let's add a test program that uses OxCaml syntax.

::::{dropdown} `main.ml`
:icon: file-code

:::{literalinclude} oxcaml/main.ml
:language: dune
:::

::::

And to build it, we add a `dune` file.

::::{dropdown} `dune`
:icon: file-code

:::{literalinclude} oxcaml/dune
:language: dune
:::

::::

## Building the project

As usual, let's lock the dependencies and build the project with:

```
$ dune pkg lock
Solution for dune.lock:                                                        
- conf-autoconf.0.2
- conf-which.1

> - ocaml-variants.5.2.0+ox

Reviewer 2: ```suggestion
```

> $ dune build && dune exec ./main.exe

Reviewer: The first step isn't necessary as the `dune exec` will build all that is needed.

...
43
```

> We have successfully built the project with the OxCaml compiler.

Reviewer 2: ```suggestion
We have successfully built the project with the OxCaml compiler and run the executable.
```
